File: be/models/voucher.model.js
const db = require("../configs/db");

const voucherSchema = new db.mongoose.Schema(
  {
    shop_id: {
      type: db.mongoose.Schema.Types.ObjectId,
      ref: "shopModel",
      default: null,
    },
    voucher_code: { type: String, required: true, unique: true },
    voucher_name: { type: String, required: true },
    voucher_type: {
      type: String,
      required: true,
      enum: ["platform", "shop"],
    },
    discount_type: {
      type: String,
      required: true,
      enum: ["percent", "amount", "free_shipping"],
    },
    discount_value: { type: Number, required: true },
    min_order_amount: { type: Number, required: true, default: 0 },
    max_discount_amount: { type: Number, default: null },
    usage_limit: { type: Number, default: null },
    used_count: { type: Number, required: true, default: 0 },
    start_date: { type: Date, required: true },
    end_date: { type: Date, required: true },
    status: {
      type: String,
      required: true,
      enum: ["active", "inactive", "expired"],
      default: "active",
    },
    deleted_at: { type: Date, default: null },
  },
  { collection: "vouchers", timestamps: true },
);

let voucherModel = db.mongoose.model("voucherModel", voucherSchema);
module.exports = { voucherModel };

File: tuhubread/lib/models/voucher.model.dart
import 'package:json_annotation/json_annotation.dart';

part 'voucher.model.g.dart';

@JsonSerializable(fieldRename: FieldRename.snake)
class VoucherModel {
  @JsonKey(name: '_id')
  final String id;

  final String? shopId;
  final String voucherCode;
  final String voucherName;
  final String voucherType;      // platform / shop
  final String discountType;     // percent / amount / free_shipping
  final double discountValue;
  final double minOrderAmount;
  final double? maxDiscountAmount;

  // Claim (save) fields
  final int? claimLimit;         // null = không giới hạn người save
  @JsonKey(defaultValue: 0)
  final int claimedCount;        // số người đã save

  // Usage fields
  final int? usageLimit;
  @JsonKey(defaultValue: 0)
  final int usedCount;

  final DateTime startDate;
  final DateTime endDate;
  final String status;           // active / inactive / expired

  VoucherModel({
    required this.id,
    this.shopId,
    required this.voucherCode,
    required this.voucherName,
    required this.voucherType,
    required this.discountType,
    required this.discountValue,
    required this.minOrderAmount,
    this.maxDiscountAmount,
    this.claimLimit,
    required this.claimedCount,
    this.usageLimit,
    required this.usedCount,
    required this.startDate,
    required this.endDate,
    required this.status,
  });

  /// Số mã còn lại có thể save. Null = không giới hạn.
  int? get remainingClaims =>
      claimLimit == null ? null : claimLimit! - claimedCount;

  /// Voucher còn mã để save không?
  bool get canBeClaimed =>
      status == 'active' &&
      endDate.isAfter(DateTime.now()) &&
      (claimLimit == null || claimedCount < claimLimit!);

  factory VoucherModel.fromJson(Map<String, dynamic> json) =>
      _$VoucherModelFromJson(json);

  Map<String, dynamic> toJson() => _$VoucherModelToJson(this);
}
